Fixed #32178 -- Allowed database backends to skip tests and mark expected failures.

hramezani opened the pull request:
https://code.djangoproject.com/ticket/32178

timgraham commented on django/db/backends/base/creation.py:
```diff
@@ -47,6 +67,8 @@ def create_test_db(self, verbosity=1, autoclobber=False, serialize=True, keepdb=
                 self._get_database_display_str(verbosity, test_database_name),
             ))
 
+        self.mark_tests()
```
This should only be called when the Django test suite is running. Perhaps with an environment variable which `runtests.py` could set, unless we can come up with a more elegant solution.

timgraham commented on django/db/backends/base/creation.py:
```diff
@@ -27,6 +32,21 @@ def _nodb_cursor(self):
         sys.stderr.write(msg + os.linesep)
 
+    def mark_tests(self):
+        tests_map = (
```
It might work to reuse the code but currently there's the drawback that a "skip reason" isn't set.

timgraham reviewed (commented):
This is on the right track. See also my WIP at the first commit of https://github.com/cockroachdb/django-cockroachdb/pull/194. It's not highly polished.

timgraham commented:
Looking good. One problem I found while working on my implementation is that I'm not sure if `@cached_property` supports inheritance. The GIS backends, for example, may need to add some skips like `geogapp/tests.py:    @skipIf(spatialite, "SpatiaLite doesn't support distance lookups with Distance objects.")`.

hramezani commented:
@timgraham It seems `@cached_property` works fine. I override the `skip_tests` in `backends/spatialite/features.py ` and remove `skipIf` from `test02_distance_lookup` in `geogapp/tests.py` and it works as expected

timgraham reviewed (commented):
You could do a careful audit of `connection.vendor` usage in tests and see if any more expected failures / skips could be added. I found a few at first glance:
https://github.com/django/django/blob/b9fd09d21b722c3745501151f9cbd5b95ccc3177/tests/annotations/tests.py#L23-L33
https://github.com/django/django/blob/b9fd09d21b722c3745501151f9cbd5b95ccc3177/tests/aggregation/tests.py#L1221-L1225

Reflecting on the second example, I wonder if we should move what's currently as comments in`skip_tests` to the datastructure so we can use it as the skip reason. That would be quite a bit more helpful than just "unsupported by <database>" which would require having to look up the test in `features.py` and finding the comment above the test.

I'm not in love with this but something like:
```python
skip_tests = {
    'GROUP BY optimization does not work properly when...' : {
        'aggreation.tests.AggregateTestCase.test_aggregation_subquery_annotation_multivalued',
        ...
    },
    # in other words
    'skip reason': {
        <set of tests>
    }
    ...
}
```

hramezani commented on django/db/backends/base/creation.py:
```diff
@@ -47,6 +71,9 @@ def create_test_db(self, verbosity=1, autoclobber=False, serialize=True, keepdb=
                 self._get_database_display_str(verbosity, test_database_name),
             ))
 
+        if os.environ.get('DJANGO_TEST_SUITE') == '1':
```
Do you mean that we should move the `connection.creation.mark_expected_failures_and_skips()` under this block in `runtests.py`?

https://github.com/django/django/blob/55fabc53373a8c7ef31d8c4cffd2a07be0a88c2e/tests/runtests.py#L299-L300

timgraham replied:
I'm not sure where exactly in runtests.py (if you have trouble with it, I'll look it at, hopefully it's feasible) but no, not there. The logic should run regardless of what test runner is used. The idea is that `runtests.py` is used for running Django's tests suite. That's all we care about. The test runner that's used doesn't matter. Jenkins, for example, uses  `TEST_RUNNER = 'xmlrunner.extra.djangotestrunner.XMLTestRunner'`. We still need the logic to run in that case.

hramezani replied:
I moved the `connection.creation.mark_expected_failures_and_skips()` to `tests.runtests.django_tests` function.
It works with django test runner and `django-docker-box`.

hramezani commented on tests/backends/base/test_creation.py:
```diff
@@ -107,6 +108,18 @@ def test_migrate_test_setting_true(self, mocked_sync_apps, mocked_migrate, *mock
             with mock.patch.object(creation, '_destroy_test_db'):
                 creation.destroy_test_db(old_database_name, verbosity=0)
 
+    @mock.patch.dict(os.environ, {"DJANGO_TEST_SUITE": "0"})
```
I removed this test

hramezani commented:
Thanks @timgraham for the review.
I found more tests that should be skipped and moved them to `django_test_skips`.

I tried to move `test_closes_connections`
https://github.com/django/django/blob/adb40d217ec57ade46b1394cfbf3c513dc669445/tests/servers/test_liveserverthread.py#L14-L18

But, the settings dict is not initialized at the beginning of the test running, and `self.connection.is_in_memory_db()` returns `False` when I moved the test to `sqlite3.features.django_test_skips`

hramezani commented:
@timgraham Could you please take another look? please let me know if I missed something. Thanks

timgraham commented:
buildbot, test on oracle.

timgraham commented:
Thanks. I pushed some mostly cosmetic edits. The issue with `test_closes_connections` might indicate that `mark_expected_failures_and_skips()` should be called later in the testing process (which probably means it couldn't happen in `runtests.py`). At this point, I think we can have someone else take a look and give input.

hramezani commented:
Thanks @timgraham. I unchecked `Patch needs improvement`

felixxm commented on django/db/backends/oracle/features.py:
```diff
@@ -72,6 +72,28 @@ class DatabaseFeatures(BaseDatabaseFeatures):
+        "Oracle requires ORDER BY in row_number, ANSI:SQL doesn't.": {
+            'expressions_window.tests.WindowFunctionTests.test_row_number_no_ordering',
+        },
+        "Raises ORA-00600: internal error code on Oracle 18.": {
```
Single quotes.

felixxm commented on django/db/backends/mysql/features.py:
```diff
@@ -50,6 +50,52 @@ class DatabaseFeatures(BaseDatabaseFeatures):
+                'model_fields.test_charfield.TestCharField.test_emoji',
+            },
+        }
+        if 'ONLY_FULL_GROUP_BY' in self.connection.sql_mode:
+            skips.update({
+                'GROUP BY optimization does not work properly when '
+                'ONLY_FULL_GROUP_BY mode is enabled on MySQL, see #31331.': {
+                    'aggregation.tests.AggregateTestCase.test_aggregation_subquery_annotation_multivalued',
+                    'annotations.tests.NonAggregateAnnotationTestCase.test_annotation_aggregate_with_m2o',
+                },
+            })
+        if (
+            self.connection.mysql_is_mariadb and
+            (10, 4, 3) < self.connection.mysql_version < (10, 5, 2)
+        ):
+            skips.update({
+                'https://jira.mariadb.org/browse/MDEV-19598': {
+                    'schema.tests.SchemaTests.test_alter_not_unique_field_to_primary_key',
+                },
+            })
+        if (
+            self.connection.mysql_is_mariadb and
+            (10, 4, 12) < self.connection.mysql_version < (10, 5)
+        ):
+            skips.update({
+                'https://jira.mariadb.org/browse/MDEV-22775': {
+                    'schema.tests.SchemaTests.test_alter_pk_with_self_referential_field',
+                },
+            })
+        if not self.supports_explain_analyze:
+            skips.update({
+                'MariaDB and MySQL >= 8.0.18 specific.': {
+                    'queries.test_explain.ExplainTests.test_mysql_analyze',
+                },
+            })
```
We cannot move these skips, because that will cause running queries before a test database is created, see 493b26bbfc9cc0ad223ece131741cba2312ced0f (ticket-31888).

timgraham replied:
Based on this and other cases, I think we should instead try to mark skips after the database is created (https://github.com/django/django/pull/13717#issuecomment-739609140).

hramezani replied:
I moved `mark_expected_failures_and_skips` function call to the end of `db.backends.base.creation.create_test_db`.
Also, I moved `servers.test_liveserverthread.LiveServerThreadTest.test_closes_connections` to skip tests of sqlite3

timgraham replied:
Assuming this works, don't forget that we'll need some solution (like the environment variable used before) to only run `mark_expected_failures_and_skips()` when running Django's test suite.

hramezani replied:
I added the environment variable

felixxm commented on django/contrib/gis/db/backends/mysql/features.py:
```diff
@@ -24,3 +24,14 @@ def supports_empty_geometry_collection(self):
+    @cached_property
+    def django_test_skips(self):
+        skips = super().django_test_skips
+        if not self.connection.mysql_is_mariadb and self.connection.mysql_version < (8, 0, 0):
+            skips.update({
+                'MySQL < 8 gives different results.': {
+                    'gis_tests.geoapp.tests.GeoLookupTest.test_disjoint_lookup',
+                },
+            })
```
This must be an inner skip, please revert (ticket-31888).

felixxm commented on tests/backends/base/test_creation.py:
```diff
@@ -188,3 +188,45 @@ def test_serialize_db_to_string_base_manager(self):
             data = connection.creation.serialize_db_to_string()
         self.assertIn('"model": "backends.schoolclass"', data)
         self.assertIn('"year": 1000', data)
+
```
I'm not sure if it's the best place for these tests :thinking:, maybe `backends/base/test_base.py`.

timgraham replied:
It's here because `mark_expected_failures_and_skips()` is on `DatabaseCreation` (`creation.py`). `test_base.py` is for `DatabaseWrapper` (`base.py`).

felixxm commented on tests/backends/base/test_creation.py:
```diff
@@ -188,3 +188,45 @@ def test_serialize_db_to_string_base_manager(self):
+            },
+        }
+        creation.mark_expected_failures_and_skips()
+        self.assertTrue(expected_failure_test_function.__unittest_expecting_failure__)
```
```suggestion
        self.assertIs(expected_failure_test_function.__unittest_expecting_failure__, True)
```

felixxm reviewed (commented):
Thanks both :rocket: I left comments.

felixxm commented:
Also, I think we need a release notes.

timgraham commented:
I understand release notes could be helpful. On the other hand, it seems odd to mention backend features in release notes without formally documenting them elsewhere.

felixxm commented:
> I understand release notes could be helpful. On the other hand, it seems odd to mention backend features in release notes without formally documenting them elsewhere.

True, but I think it's still valuable for maintainers of 3rd-party database backends :thinking:

felixxm commented on tests/backends/base/test_creation.py:
```diff
@@ -188,3 +188,45 @@ def test_serialize_db_to_string_base_manager(self):
+            SkipTestClass.__unittest_skip_why__,
+            'skip test class',
+        )
+        self.assertTrue(skip_test_function.__unittest_skip__)
```
```suggestion
        self.assertIs(skip_test_function.__unittest_skip__, True)
```

hramezani commented:
@felixxm  release note added

timgraham commented on docs/releases/3.2.txt:
```diff
@@ -451,6 +451,17 @@ Tests
   comparison against a list of string representations of objects when using the
   default value for the ``transform`` argument.
 
+* Database backends now can mark tests as expected failure by adding them to
```
Please put this in a new "Database backends" section instead.

* Third-party database backends can now skip or mark as expected failures tests in Django's test suite using the new DatabaseFeatures.django_test_skips and django_test_expected_failures attributes.

The documentation isn't needed because authors should consult the existing backends to see usage.

timgraham commented on django/db/backends/base/creation.py:
```diff
@@ -94,7 +94,8 @@ def create_test_db(self, verbosity=1, autoclobber=False, serialize=True, keepdb=
         self.connection.ensure_connection()
 
-        self.mark_expected_failures_and_skips()
+        if os.environ.get('DJANGO_TEST_SUITE') == "1":
```
single quotes

timgraham commented on tests/runtests.py:
```diff
@@ -248,6 +248,8 @@ def _module_match_label(module_label, label):
 
     apps.set_installed_apps(settings.INSTALLED_APPS)
 
+    os.environ['DJANGO_TEST_SUITE'] = "1"
```
single quotes + `RUNNING_DJANGOS_TEST_SUITE` could be more descriptive. Also comment `# Set an environment variable that other code may consult to see if Django's own test suite is running.`

felixxm commented:
Thanks :1st_place_medal:

felixxm commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.

timgraham commented on django/db/backends/sqlite3/features.py:
```diff
@@ -50,6 +50,38 @@ class DatabaseFeatures(BaseDatabaseFeatures):
+                    'servers.test_liveserverthread.LiveServerThreadTest.test_closes_connections',
+                },
+            })
+
```
chop

timgraham commented on tests/backends/base/test_creation.py:
```diff
@@ -107,6 +108,18 @@ def test_migrate_test_setting_true(self, mocked_sync_apps, mocked_migrate, *mock
+    @mock.patch.dict(os.environ, {'RUNNING_DJANGOS_TEST_SUITE': ''})
+    @mock.patch.object(BaseDatabaseCreation, 'mark_expected_failures_and_skips')
+    def test_mark_expected_failures_and_skips_call(self, mark_expected_failures_and_skips, *mocked_objects):
+        test_connection = get_connection_copy()
```
A docstring could be nice: "mark_expected_failures_and_skips() isn't called unless RUNNING_DJANGOS_TEST_SUITE is 'true'."

felixxm commented:
buildbot

timgraham commented on django/db/backends/base/features.py:
```diff
@@ -323,6 +323,13 @@ class BaseDatabaseFeatures:
 
+    # A set of dotted paths to tests in Django's test suite that are expected
+    # to fail on this database.
+    django_test_expected_failures = {}
```
Oops, this should be `set()`. None of the subclasses implementations call `super()` which perhaps isn't necessary but I want the CockroachDB backend to call `super()` to get any skips/expected failures from the PostgreSQL backend  (which the CockroachDB inherits),

hramezani replied:
I can create a PR to fix it

hramezani replied:
Here is the PR https://github.com/django/django/pull/13761

felixxm replied:
Good catch :dart:
